File: 2024/day6.py
import AOCInit
import util

if __name__ != '__main__':
    exit()

inp = util.getInput(d=6, y=2024)

labMap = inp.splitlines()
n = len(labMap)
m = len(labMap[0])
dirs: tuple[tuple[int, int], ...] = ((-1, 0), (0, 1), (1, 0), (0, -1))

obs = [
    [labMap[i][j] == '#' for j in range(m)]
    for i in range(n)
]

# part 1
x0, y0 = divmod(inp.index('^'), m + 1)
x, y = x0, y0
guardDir = 0
visited = set()
while 0 <= x < n and 0 <= y < m:
    visited.add((x, y))
    xx = x + dirs[guardDir][0]
    yy = y + dirs[guardDir][1]
    rightDir = (guardDir + 1) % 4
    if 0 <= xx < n and 0 <= yy < m and obs[xx][yy]:
        guardDir = rightDir
    else:
        x, y = xx, yy
print(len(visited))

# part 2
def walkHasLoop():
    x, y = x0, y0
    d = 0
    visitedState = set()
    while 0 <= x < n and 0 <= y < m:
        if (x, y, d) in visitedState:
            return True
        visitedState.add((x, y, d))
        xx = x + dirs[d][0]
        yy = y + dirs[d][1]
        rightDir = (d + 1) % 4
        if 0 <= xx < n and 0 <= yy < m and obs[xx][yy]:
            d = rightDir
        else:
            x, y = xx, yy
    return False

counter = 0
for pt in visited:
    if pt == (x0, y0):
        continue
    obs[pt[0]][pt[1]] = True
    counter += walkHasLoop()
    obs[pt[0]][pt[1]] = False
print(counter)

File: 2024/day6.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_set>
#include <vector>

#include "../util.hpp"

using lli = long long int;

constexpr int dirs[4][2] = {
    {-1, 0}, {0, 1}, {1, 0}, {0, -1}
};

bool hasLoop(int x, int y, int n, int m, std::unordered_set<lli> &obs){
    int d = 0;
    std::unordered_set<lli> visited;
    while (util::inRange(x, 0, n) && util::inRange(y, 0, m)){
        int z = (x * m + y) * 4 + d;
        if (util::contains(z, visited)){
            return true;
        }
        visited.insert(z);
        int xx = x + dirs[d][0], yy = y + dirs[d][1];
        if (util::contains(xx * m + yy, obs)){
            d = (d + 1) % 4;
        } else {
            x = xx;
            y = yy;
        }
    }
    return false;
}

int main(int, char**){
    std::stringstream inpStream;
    util::getInput(2024, 6, inpStream);
    std::vector<std::string> labMap(util::splitline(inpStream));
    int n = labMap.size(), m = labMap[0].size();
    int x0, y0;
    std::unordered_set<lli> obs;
    for (int i = 0; i < n; ++i){
        for (int j = 0; j < m; ++j){
            if (labMap[i][j] == '^'){
                x0 = i;
                y0 = j;
            } else if (labMap[i][j] == '#'){
                obs.insert(i * m + j);
            }
        }
    }

    // part 1
    int x = x0, y = y0, d = 0;
    std::unordered_set<lli> visited;
    while (util::inRange(x, 0, n) && util::inRange(y, 0, m)){
        visited.insert(x * m + y);
        int xx = x + dirs[d][0], yy = y + dirs[d][1];
        if (util::contains(xx * m + yy, obs)){
            d = (d + 1) % 4;
        } else {
            x = xx;
            y = yy;
        }
    }
    std::cout << visited.size() << std::endl;

    // part 2
    int count = 0;
    for (lli pt : visited){
        int xx = pt / m, yy = pt % m;
        if (xx == x0 && yy == y0){
            continue;
        }
        obs.insert(xx * m + yy);
        count += hasLoop(x0, y0, n, m, obs);
        obs.erase(xx * m + yy);
    }
    std::cout << count << std::endl;

    return 0;
}
